### PR TITLE
[TACHYON-1420] Check - Check Tachyon specfic operations in Master side

### PR DESCRIPTION
**Pull request by sundapeng**
https://tachyon.atlassian.net/browse/TACHYON-1420

This pull request is adding permission check for setState, completeFile, free, loadMetadata, mount, unmount.

@apc999 , @jsimsa , @yupeng9, @dongche

### CODE REVIEW

File: servers/src/main/java/tachyon/master/file/FileSystemMaster.java [hunk 1]
```diff
@@ -1783,62 +1795,66 @@ private void setStateFromEntry(SetStateEntry entry) throws FileDoesNotExistExcep
...
+        fileInfos.add(inode.generateClientFileInfo(mInodeTree.getPath(inode).toString()));
+      }
 
-    String[] pathComponents = PathUtils.getPathComponents(path.getPath());
-    if (pathComponents.length < fileInfos.size()) {
-      throw new InvalidPathException(ExceptionMessage.PATH_INVALID.getMessage(path.getPath()));
-    }
+      String[] pathComponents = PathUtils.getPathComponents(path.getPath());
+      if (pathComponents.length < fileInfos.size()) {
+        throw new InvalidPathException(ExceptionMessage.PATH_INVALID.getMessage(path.getPath()));
```
Comment (jsimsa) [inline servers/src/main/java/tachyon/master/file/FileSystemMaster.java]: can any other part of the try block throw the InvalidPathException? if not, remove the try block and just log the warning here and return

File: servers/src/main/java/tachyon/master/file/FileSystemMaster.java [hunk 2]
```diff
@@ -1366,6 +1369,7 @@ public long loadMetadata(TachyonURI path, boolean recursive)
         throw new FileDoesNotExistException(
             ExceptionMessage.PATH_DOES_NOT_EXIST.getMessage(path.getPath()));
       }
+      checkPermission(FileSystemAction.READ, path, false);
```
Comment (jsimsa) [inline servers/src/main/java/tachyon/master/file/FileSystemMaster.java]: why do we need this here?

File: servers/src/main/java/tachyon/master/lineage/recompute/RecomputeExecutor.java [hunk 3]
```diff
@@ -93,6 +94,8 @@ public void run() {
               mFileSystemMaster.resetFile(fileId);
             } catch (FileDoesNotExistException e) {
               LOG.error("the lost file {} is invalid", fileId, e);
+            } catch (AccessControlException e) {
+              LOG.error("the file {} is unauthorized", fileId, e);
```
Comment (jsimsa) [inline servers/src/main/java/tachyon/master/lineage/recompute/RecomputeExecutor.java]: the lost file {} cannot be accessed

File: servers/src/test/java/tachyon/master/file/FileSystemMasterPermissionCheckTest.java [hunk 4]
```diff
@@ -468,6 +471,174 @@ public void readDirFailTest() throws Exception {
...
+    verifyCreate(TEST_USER_1, file, false);
+    SetStateOptions expect = getRamdomSetState();
+
+    mThrown.expect(AccessControlException.class);
+    mThrown.expectMessage(ExceptionMessage.PERMISSION_DENIED.getMessage(toExceptionMessage(
+        TEST_USER_2.getUser(), FileSystemAction.WRITE, file, "testState1")));
+    verifySetState(TEST_USER_2, getFileId(TEST_USER_1, file), expect);
+  }
+
+  private SetStateOptions getRamdomSetState() {
```
Comment (jsimsa) [inline servers/src/test/java/tachyon/master/file/FileSystemMasterPermissionCheckTest.java]: Random

File: servers/src/test/java/tachyon/master/file/FileSystemMasterPermissionCheckTest.java [hunk 5]
```diff
@@ -468,6 +471,174 @@ public void readDirFailTest() throws Exception {
...
+    conf.set(Constants.SECURITY_AUTHORIZATION_PERMISSIONS_UMASK, "044");
+    MasterContext.reset(conf);
+
+    String file = TEST_DIR_URI + "/testState1";
+    verifyCreate(TEST_USER_1, file, false);
+    verifyFree(TEST_USER_2, getFileId(TEST_USER_1, file), false);
+  }
+
+  @Test
+  public void freeNonNullFolderSuccessTest() throws Exception {
```
Comment (jsimsa) [inline servers/src/test/java/tachyon/master/file/FileSystemMasterPermissionCheckTest.java]: s/Folder/Directory

File: servers/src/test/java/tachyon/master/file/FileSystemMasterPermissionCheckTest.java [hunk 6]
```diff
@@ -468,6 +471,174 @@ public void readDirFailTest() throws Exception {
...
+    verifyCreate(TEST_USER_1, file, false);
+
+    mThrown.expect(AccessControlException.class);
+    mThrown.expectMessage(ExceptionMessage.PERMISSION_DENIED.getMessage(toExceptionMessage(
+        TEST_USER_2.getUser(), FileSystemAction.WRITE, file, "testComplete1")));
+    verifyFree(TEST_USER_2, getFileId(TEST_USER_1, file), false);
+  }
+
+  @Test
+  public void freeNonNullFolderFailTest() throws Exception {
```
Comment (jsimsa) [inline servers/src/test/java/tachyon/master/file/FileSystemMasterPermissionCheckTest.java]: s/Folder/Directory

**Comment by jsimsa** [comment]
@sundapeng thanks for the PR ... I have one high-level comment ... currently, checking of the permission is "sprinkled" throughout the master code base and for some operations that can throws the AccessControlException, the permission checking is delegated to some method they call ... this makes the code hard to read ... I would argue that each method that can throw AccessControlException, should check the permission as soon as possible, ideally as the first thing that the method does and fail early ... I think this would establish an easy to follow permission checking convention ... instead of expecting that the developer modifying the existing code understand which methods do the actually checking (e.g. imagine someone would change resetFile() to longer call free() without realizing that free() was actually doing the permission checking)

let me know what you think and if you agree ... let's update this PR so that all functions that can fail because of access control, fail early

File: servers/src/test/java/tachyon/master/file/FileSystemMasterPermissionCheckTest.java [hunk 7]
```diff
@@ -468,6 +471,174 @@ public void readDirFailTest() throws Exception {
...
+  }
+
+  @Test
+  public void setStateFailTest() throws Exception {
+    // set unmask
+    TachyonConf conf = MasterContext.getConf();
+    conf.set(Constants.SECURITY_AUTHORIZATION_PERMISSIONS_UMASK, "066");
+    MasterContext.reset(conf);
+
+    String file = TEST_DIR_URI + "/testState1";
```
Comment (apc999) [inline servers/src/test/java/tachyon/master/file/FileSystemMasterPermissionCheckTest.java]: `PathUtils.concatPath` please, for better portability.

File: servers/src/test/java/tachyon/master/file/FileSystemMasterPermissionCheckTest.java [hunk 8]
```diff
@@ -468,6 +471,174 @@ public void readDirFailTest() throws Exception {
...
+  @Test
+  public void setStateSuccessTest() throws Exception {
+    // set unmask
+    TachyonConf conf = MasterContext.getConf();
+    conf.set(Constants.SECURITY_AUTHORIZATION_PERMISSIONS_UMASK, "044");
+    MasterContext.reset(conf);
+
+    String file = TEST_DIR_URI + "/testState1";
+    verifyCreate(TEST_USER_1, file, false);
+    SetStateOptions expect = getRamdomSetState();
```
Comment (apc999) [inline servers/src/test/java/tachyon/master/file/FileSystemMasterPermissionCheckTest.java]: in general, we avoid introduce randomness in unit tests, as test failures are not surely reproducible. Can you simply pick ttl/recursive flag here as they seem not critical here

File: servers/src/main/java/tachyon/master/file/FileSystemMaster.java (hunk 1 above)
Comment (apc999) [inline servers/src/main/java/tachyon/master/file/FileSystemMaster.java]: I think checkParentPermission throws this exception below

File: servers/src/main/java/tachyon/master/file/FileSystemMaster.java [hunk 9]
```diff
@@ -1783,62 +1795,66 @@ private void setStateFromEntry(SetStateEntry entry) throws FileDoesNotExistExcep
...
+      for (Inode inode : inodes) {
+        fileInfos.add(inode.generateClientFileInfo(mInodeTree.getPath(inode).toString()));
+      }
 
-    String[] pathComponents = PathUtils.getPathComponents(path.getPath());
-    if (pathComponents.length < fileInfos.size()) {
-      throw new InvalidPathException(ExceptionMessage.PATH_INVALID.getMessage(path.getPath()));
-    }
+      String[] pathComponents = PathUtils.getPathComponents(path.getPath());
+      if (pathComponents.length < fileInfos.size()) {
```
Comment (apc999) [inline servers/src/main/java/tachyon/master/file/FileSystemMaster.java]: why `pathComponents.length` could be greater than `fileInfos.size()`?

Comment (apc999) [inline servers/src/main/java/tachyon/master/file/FileSystemMaster.java]: are we here concerned the case where for path '/a/b/c/', only '/a' is in inode tree, while 'b/c' is not?

**Comment by apc999** [comment]
@sundapeng  thanks for the PR! I left a few questions as well as comments.

**Comment by yupeng9** [comment]
I agree with @jsimsa. In fact, it'll be cleaner if the permission checking is outside `FileSystemMaster`. The checker should serve as a gatekeeper before calling the methods on `FileSystemMaster`, so maybe we can put it in the service handler or a wrapper of `FileSystemMaster`?

File: servers/src/main/java/tachyon/master/file/FileSystemMaster.java (hunk 1 above)
Comment (dongche) [inline servers/src/main/java/tachyon/master/file/FileSystemMaster.java]: Yes, that method also throws this exception. It needs to ensure the path is valid before checking the path

File: servers/src/main/java/tachyon/master/file/FileSystemMaster.java (hunk 9 above)
Comment (dongche) [inline servers/src/main/java/tachyon/master/file/FileSystemMaster.java]: Normally `pathComponents.length` is not possible to be greater than `fileInfos.size()`. Just add this for full coverage of all case. So it can be removed.

File: servers/src/main/java/tachyon/master/file/FileSystemMaster.java (hunk 1 above)
Comment (sundapeng) [inline servers/src/main/java/tachyon/master/file/FileSystemMaster.java]: Okay, I will keep the try-catch block and changed the exception here to log the warning and return.

File: servers/src/main/java/tachyon/master/lineage/recompute/RecomputeExecutor.java (hunk 3 above)
Comment (sundapeng) [inline servers/src/main/java/tachyon/master/lineage/recompute/RecomputeExecutor.java]: Okay, I will fix it.

File: servers/src/test/java/tachyon/master/file/FileSystemMasterPermissionCheckTest.java (hunk 4 above)
Comment (sundapeng) [inline servers/src/test/java/tachyon/master/file/FileSystemMasterPermissionCheckTest.java]: Okay, I will fix it.

File: servers/src/test/java/tachyon/master/file/FileSystemMasterPermissionCheckTest.java (hunk 5 above)
Comment (sundapeng) [inline servers/src/test/java/tachyon/master/file/FileSystemMasterPermissionCheckTest.java]: Okay, I will fix it.

File: servers/src/test/java/tachyon/master/file/FileSystemMasterPermissionCheckTest.java (hunk 6 above)
Comment (sundapeng) [inline servers/src/test/java/tachyon/master/file/FileSystemMasterPermissionCheckTest.java]: Okay, I will fix it.

File: servers/src/test/java/tachyon/master/file/FileSystemMasterPermissionCheckTest.java (hunk 8 above)
Comment (sundapeng) [inline servers/src/test/java/tachyon/master/file/FileSystemMasterPermissionCheckTest.java]: Thank you, I will fix it.

File: servers/src/test/java/tachyon/master/file/FileSystemMasterPermissionCheckTest.java (hunk 7 above)
Comment (sundapeng) [inline servers/src/test/java/tachyon/master/file/FileSystemMasterPermissionCheckTest.java]: Okay, I will fix it.

File: servers/src/main/java/tachyon/master/file/FileSystemMaster.java (hunk 2 above)
Comment (sundapeng) [inline servers/src/main/java/tachyon/master/file/FileSystemMaster.java]: yes, it should move before **ufsPath = mMountTable.resolve(path);**

**Comment by sundapeng** [comment]
@jsimsa Do you mean adding permission check in each method that can throw AccessControlException, instead of delegating to other operation.
But this will cause check logic to be executed twice in some case. For you example, resetFile() calls free() and they may both do same check. They are both public and interleave together. If this is ok, we can update in next patch.
@yupeng9 About the solution of `wrapper`, some `checkPermission`s are in a same `synchronized` with  operation, if we bring the `checkPermission` out of `synchronized`, it may cause transaction issue.

**Comment by sundapeng** [comment]
Jenkins, test this please.

**Comment by jsimsa** [comment]
@sundapeng I think what @yupeng9 is proposing is to create a separate access control "layer", e.g. called FileSystemMasterAccessControl. Its methods are invoked from FileSystemMasterClientServiceHandler and it in turns invokes FileSystemMaster methods. Thus, permission checking will be done just once at the beginning of each RPC and the nice aspect of this implementation is that it makes permission checking modular and leaves FileSystemMaster untouched.

File system master is currently single threaded so it is fine for the FileSystemMasterAccessControl to grab a lock on the Inode tree any time it wants to check permission to "set and check" data races.

**Comment by sundapeng** [comment]
Hi @jsimsa and @yupeng9 
About the refactory, since https://github.com/amplab/tachyon/pull/2369  which have same permission check have been merged, it may be not very suitable do the refactory at this pr. We also need to consider the issue likes granularity of the lock for the refactory.
We have discussed it with @apc999 on today's online discussion. Since https://github.com/amplab/tachyon/pull/2369 is merged, we think it is better to make the function is integrated first. Do you think it is possible to merge this pr before the refactory?

**Comment by yupeng9** [comment]
Thanks, @sundapeng. That sounds reasonable to me. Could you add another ticket and do it in a follow-up PR?

**Comment by sundapeng** [comment]
Okay, filed https://tachyon.atlassian.net/browse/TACHYON-1517 to tracking it.

File: servers/src/test/java/tachyon/master/file/FileSystemMasterPermissionCheckTest.java [hunk 10]
```diff
@@ -468,6 +471,172 @@ public void readDirFailTest() throws Exception {
...
+  }
+
+  @Test
+  public void freeNonNullDirectorySuccessTest() throws Exception {
+    // set unmask
+    TachyonConf conf = MasterContext.getConf();
+    conf.set(Constants.SECURITY_AUTHORIZATION_PERMISSIONS_UMASK, "044");
+    MasterContext.reset(conf);
+
+    String subFolder = PathUtils.concatPath(TEST_DIR_URI, "/testState");
```
Comment (jsimsa) [inline servers/src/test/java/tachyon/master/file/FileSystemMasterPermissionCheckTest.java]: nit: s/subFolder/subDir

**Comment by jsimsa** [comment]
just one minor comment, otherwise LGTM

File: servers/src/test/java/tachyon/master/file/FileSystemMasterPermissionCheckTest.java (hunk 10 above)
Comment (sundapeng) [inline servers/src/test/java/tachyon/master/file/FileSystemMasterPermissionCheckTest.java]: Thank you, I will fix it.

**Comment by jsimsa** [comment]
LGTM

File: servers/src/main/java/tachyon/master/file/FileSystemMaster.java [hunk 11]
```diff
@@ -1783,65 +1795,72 @@ private void setStateFromEntry(SetStateEntry entry) throws FileDoesNotExistExcep
   }
 
   /**
-   * Checks user's permission on a path.
+   * Checks user's permission on a path. If the path is invalid, it should bypass the
+   * {@link InvalidPathException} and the logic at operation will handle it.
```
Comment (yupeng9) [inline servers/src/main/java/tachyon/master/file/FileSystemMaster.java]: Could you add a comment that the caller should lock `mInodeTree` before calling it?

**Comment by yupeng9** [comment]
@sundapeng I have one more comment, other than that, it LGTM

**Comment by sundapeng** [comment]
Thank @yupeng9  @jsimsa for your review.

**Comment by sundapeng** [comment]
Jenkins, test this please.

**Comment by apc999** [comment]
+1 LGTM

File: servers/src/test/java/tachyon/master/file/FileSystemMasterTest.java [hunk 12]
```diff
@@ -21,6 +21,9 @@
 import java.util.concurrent.Future;
 import java.util.concurrent.TimeUnit;
 
+import com.google.common.collect.ImmutableMap;
```
Comment (apc999) [inline servers/src/test/java/tachyon/master/file/FileSystemMasterTest.java]: This import order is inconsistent with our requirement:
you probably want to revert this change for this file (the original import order is correct).

**Comment by apc999** [comment]
@sundapeng I was LGTMing your PR, until I found an import order problem.
Could you address that? And I will then merge your PR.

Thanks!

**Comment by sundapeng** [comment]
Thank @apc999  for your comments, I have updated the patch.

**Comment by apc999** [comment]
+1 LGTM

Thanks for the patch!

**Comment by sundapeng** [comment]
Thank @apc999 @yupeng9 @jsimsa for your review.
